`cfg_simplify!` error

<!--
If you have a question please search or post to our Discourse site: https://discourse.julialang.org.
We use the GitHub issue tracker for bug reports and feature requests only.

If you're submitting a bug report, be sure to include as much relevant information as
possible, including:

1. The output of `versioninfo()`
2. How you installed Julia
3. A minimal working example (MWE), also known as a minimum reproducible example

If you're experiencing a problem with a particular package, open an issue on that
package's repository instead.

Thanks for contributing to the Julia project!
-->

MRE:
```julia
julia> ir, = Base.code_ircode() do
           while true
               i = @noinline rand(Int)
               if @noinline isodd(i)
                   return i
               end
           end
       end |> only
  1 ─      nothing::Nothing
3 2 ─ %2 = invoke Main.rand(Main.Int::Type{Int64})::Int64
4 │   %3 = invoke Main.isodd(%2::Int64)::Bool
  └──      goto #4 if not %3
5 3 ─      return %2
7 4 ─      goto #1
   => Int64                                                                                                                                                                   

julia> Core.Compiler.cfg_simplify!(ir)
ERROR: AssertionError: label > 0
Stacktrace:
 [1] 
   @ Core.Compiler ./compiler/ssair/ir.jl:1348
 [2] cfg_simplify!(ir::Core.Compiler.IRCode)
   @ Core.Compiler ./compiler/ssair/passes.jl:2276
 [3] top-level scope
   @ REPL[8]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

This issue particularly blocks #51960, which introduces actual use of `cfg_simplify!` into the base compilation pipeline. To echo what Keno mentioned in our previous Slack chat:
> Keno Fischer 
> I'd have to look into it, but the thing that strikes me as unusual about this IR is that we loop into the entry block. I wouldn't be surprised if we accidentally used the entry as a magic value somewhere and got confused

--- a/base/compiler/ssair/passes.jl
+++ b/base/compiler/ssair/passes.jl
@@ -2029,7 +2029,7 @@ function cfg_simplify!(ir::IRCode)
         i = popfirst!(worklist)
         # Drop blocks that will be merged away
         if merge_into[i] != 0
-            bb_rename_succ[i] = -1
+            bb_rename_succ[i] = typemin(Int)
         end
         # Mark dropped blocks for fixup
         if !isempty(searchsorted(dropped_bbs, i))
@@ -2049,7 +2049,7 @@ function cfg_simplify!(ir::IRCode)
                 # we have to schedule that block next
                 while merged_succ[curr] != 0
                     if bb_rename_succ[curr] == 0
-                        bb_rename_succ[curr] = -1
+                        bb_rename_succ[curr] = typemin(Int)
                     end
                     curr = merged_succ[curr]
                 end
@@ -2089,9 +2089,9 @@ function cfg_simplify!(ir::IRCode)
         resolved_all = true
         for bb in dropped_bbs
             obb = bb_rename_succ[bb]
-            if obb < -1
+            if obb < 0 && obb != typemin(Int)
                 nsucc = bb_rename_succ[-obb]
-                if nsucc == -1
+                if nsucc == typemin(Int)
                     nsucc = -merge_into[-obb]
                 end
                 bb_rename_succ[bb] = nsucc
@@ -2106,6 +2106,8 @@ function cfg_simplify!(ir::IRCode)
         if bb_rename_succ[i] == 0
             bb_rename_succ[i] = -1
             bb_rename_pred[i] = -2
+        elseif bb_rename_succ[i] == typemin(Int)
+            bb_rename_succ[i] = -1
         end
     end
 

--- a/test/compiler/irpasses.jl
+++ b/test/compiler/irpasses.jl
@@ -1569,4 +1569,19 @@ let m = Meta.@lower 1 + 1
     Core.Compiler.verify_ir(ir)
 end
 
+function f_with_merge_to_entry_block()
+    while true
+        i = @noinline rand(Int)
+        if @noinline isodd(i)
+            return i
+        end
+    end
+end
+
+let (ir, _) = only(Base.code_ircode(f_with_merge_to_entry_block))
+    Core.Compiler.verify_ir(ir)
+    ir = Core.Compiler.cfg_simplify!(ir)
+    Core.Compiler.verify_ir(ir)
+end
+
 # JET.test_opt(Core.Compiler.cfg_simplify!, (Core.Compiler.IRCode,))

